runtime/shared_heap.c

```diff
@@ -785,6 +817,407 @@ void caml_verify_heap(caml_domain_state *domain) {
   caml_stat_free(st);
 }
 
+/* Compaction start */
+
+/* Given a single value `v`, found at `p`, check if it points in to an
+   evacuated pool and if so update it using the forwarding pointer
+   created by the compactor. */
+static inline void compact_update_value(void* ignored,
+                                        value v,
+                                        volatile value* p)
+{
+  if (Is_block(v)) {
+    CAMLassert(!Is_young(v));
+
+    tag_t tag = Tag_val(v);
+
+    int infix_offset = 0;
+    if (tag == Infix_tag) {
+      infix_offset = Infix_offset_val(v);
+      /* v currently points to an Infix_tag inside of a Closure_tag.
+        The forwarding pointer we want is in the first field of the
+        Closure_tag. */
+      v -= infix_offset;
+      CAMLassert(Tag_val(v) == Closure_tag);
+    }
+
+    /* non-markable blocks can't move */
+    if (Has_status_val(v, NOT_MARKABLE))
+      return;
+
+    if (Whsize_val(v) <= SIZECLASS_MAX) {
+      if (caml_pool_of_shared_block(v)->evacuating) {
+        value fwd = Field(v, 0) + infix_offset;
+        CAMLassert(!caml_pool_of_shared_block(fwd)->evacuating);
+        CAMLassert(Is_block(fwd));
+        CAMLassert(Tag_val(fwd) == tag);
+        *p = fwd;
+      }
+    }
+  }
+}
+
+/* Given a value found at `p` check if it points in to an
+   evacuated pool and if so update it using the forwarding pointer
+   created by the compactor. */
+static inline void compact_update_value_at(volatile value* p)
+{
+  compact_update_value(NULL, *p, p);
+}
+
+/* For each pointer in the block pointed to by `p` check if it
+   points in to an evacuated pool and if so update it using the
+   forwarding pointer created by the compactor. */
+static void compact_update_block(value* p)
```

Comment:
`p` doesn't point to a value, it points to a header. For readability, I would write this:
```suggestion
static void compact_update_block(header_t * hp)
```